### ISSUE TITLE
Slow selecting of .mov files

### ISSUE THREAD

**Issue by fdcgr** [miguelpruivo/flutter_file_picker#531]
**Describe the bug**
We are currently experience a slow selecting of .mov files using this FilePicker. One issue is that OS Picker closes as soon as the file is selected but the app cannot proceed to compute the file because the async is hanging for 1 min until the function continues. During this time the previous screen of the app is shown which allows to use to continue to navigate in the app. The last issue can be fixed using the Status callback, however taking 1 min to load a file seems a bit excessive considering that the file is only 60Mb in size.

I am able to provide the video for which this is happening but cannot do it via this channel unfortunately. Is there any channel you would prefer for me to send the video?

**Issue details**
1. Platform - iOS
2. Platform OS version - 14.2
3. File type (filter) — .mov
4. Detailed steps to reproduce
    1. We have a .mov file in gallery with around 60Mb
    2. We use the FilePicker library to pick the file (using FileType.video)
    3. The OS Picker closes
    4. The flutter app hangs around for 1 min until the async function continues exection

**Flutter Version details**
```
[✓] Flutter (Channel stable, 1.22.3, on Mac OS X 10.15.7 19H15, locale en-PT)
    • Flutter version 1.22.3 at /usr/local/flutter
    • Framework revision 8874f21e79 (5 weeks ago), 2020-10-29 14:14:35 -0700
    • Engine revision a1440ca392
    • Dart version 2.10.3

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/guilhermeramos/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.2, Build version 12B45b
    • CocoaPods version 1.9.1

[✓] VS Code (version 1.28.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.21.1

[✓] Connected device (1 available)
    • iPhone 12 Pro Max (mobile) • 27EAD690-9C3D-4258-A2D2-98A7FC4143B7 • ios • com.apple.CoreSimulator.SimRuntime.iOS-14-2 (simulator)
```

**Comment by miguelpruivo** [comment]
Hi @fdcgr this looks like an OS limitation. The picker delegates the picked file immediately to the underlying platform. If it takes long to load, that’s probably because tb the OS is also taking long to provide it to the plugin code.

Feel free to debug it yourself and check if there’s anything that can be done in order to improve it.

Thank you.

**Comment by fdcgr** [comment]
Hi @miguelpruivo thank you for the answer. Is the OS giving us any information that could be later used to display a progress bar into the loading of the file?

Cheers,

**Comment by miguelpruivo** [comment]
Have you managed to solve it?

**Comment by fdcgr** [comment]
I tried the same file with the image_picker plugin and that apparently works. Regarding the issue, if there is none we could do on the library side I guess we could close this.

**Comment by miguelpruivo** [comment]
Hum you’ve mentioned that you had speed issues selecting mov files, not actually related with an issue selecting it.

If you use single pick on file_picker, it acts just the same as image_picker (using native API) whereas with multi it actually uses a 3rd party lib. Might be related.

**Comment by fdcgr** [comment]
I am using the single pick on file_picker.

**Comment by miguelpruivo** [comment]
Then it should take barely the same time. Are you using `withData` set to true? If so, that might explain.

**Comment by fdcgr** [comment]
I am using these parameters:
```
bool allowCompression = true,
bool allowMultiple = false,
bool withData = false,
```

**Comment by miguelpruivo** [comment]
That's weird, then it should behave the same way as `image_picker` then. If you could provide me with more details on what could be causing the slowness, that would be great, since you've tried both.

Thank you!

**Comment by fdcgr** [comment]
Hi @miguelpruivo.

I did some more debugging and was able to find a couple of more things:
- If I select just a recorded video from the device, the problem still happens but it is way faster. Logs in the xcode (for both videos) are the following:
```
2020-12-10 11:37:48.463885+0000 Runner[5153:2550061] 

***** Picker:<PHPickerViewController: 0x11d36aa10> didFinishPicking:(
    "<PHPickerResult: 0x2833f9080>"
)
* -[FilePickerPlugin picker:didFinishPicking:] [Line 365]
2020-12-10 11:37:48.466935+0000 Runner[5153:2551258] [default] [ERROR] Could not create a bookmark: NSError: Cocoa 257 "The file couldn’t be opened because you don’t have permission to view it." }
2020-12-10 11:37:56.020103+0000 Runner[5153:2550938] TX and TY are 0. Rotation: 0. Natural width,height: 1920.000000, 1080.000000
```

I also tried a different approach: Show a processing screen with a cancelation button. I tried the follow: start a file pick, cancel in the middle of processing and start to pick a new file. This also leads to a error case because the background process for picking the file is still in progress. Here are the logs for that:
```
2020-12-10 11:08:30.986563+0000 Runner[5100:2539867] flutter: [MethodChannelFilePicker] Platform exception: PlatformException(multiple_request, Cancelled by a second request, null, null)
2020-12-10 11:08:30.987208+0000 Runner[5100:2539867] flutter: ----------------FIREBASE CRASHLYTICS----------------
2020-12-10 11:08:30.987271+0000 Runner[5100:2539867] flutter: PlatformException(multiple_request, Cancelled by a second request, null, null)
2020-12-10 11:08:30.987451+0000 Runner[5100:2539867] flutter: 
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:582:7)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:159:18)
<asynchronous suspension>
#2      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:332:12)
#3      MethodChannel.invokeListMethod (package:flutter/src/services/platform_channel.dart:345:41)
#4      FilePickerIO._getPath (package:file_picker/src/file_picker_io.dart:83:47)
#5      FilePickerIO.pickFiles (package:file_picker/src/file_picker_io.dart:31:7)
#6      FilePicker.getFile (package:project_common/src/utils/file_picker.dart:20:49)
#7      FilePicker.getVideoFile (package:project_common/src/utils/file_picker.dart:46:41)
#8      _SelectVideoState._pickVideo (package:project_common/src/screens/courses/select_video.dart:43:36)
#9      _SelectVideoState.pickAction (package:project_common/src/screens/courses/select_video.dart:60:7)
#10     new Future.microtask.<anonymous closure> (dart:async/future.dart:201:37)
#11     _rootRun (dart:async/zone.dart:1182:47)
#12     _CustomZone.run (dart:async/zone.dart:1093:19)
#13     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
#14     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
#15     _rootRun (dart:async/zone.dart:1190:13)
#16     _CustomZone.run (dart:async/zone.dart:1093:19)
#17     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
#18     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
#19     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#20     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
2020-12-10 11:08:30.988677+0000 Runner[5100:2539867] flutter: ----------------------------------------------------
```

**Comment by miguelpruivo** [comment]
Thank you for further investigating this, however, I don't see an issue with it. Most of the times, if files are stored on iCloud or are big files (for example) need to be loaded asynchronously, hence, a proper way to cancel it is by invoking the picker again (for a new request) which will discard the previous one.

Let me know if you come up with a better solution/alternative for it. I'd like to make it faster nevertheless (if possible).

**Comment by fdcgr** [comment]
Isn't it strange that the same file does not lead to the same effect using the image_picker plugin? Also the file is locally store so it would not need to be downloaded from iCloud.

Regardless, thank you for the help, the last tip helped me fix the cancelation process.

**Comment by miguelpruivo** [comment]
I believe that `image_picker` loads the file synchronously, which has its pros and cons — for most users they rather have it async so it won't block the main thread while loading the file, however, the downside is exactly what you are experiencing. Sometimes it may take a bit more and you have to handle the status.

**Comment by fdcgr** [comment]
Ok, that makes sense. I guess that regarding this point I would consider fixed! As a further point, having the progress in the progress status would be great to improve our UIs but that is out of scope.

**Comment by SubashManian** [comment]
I think this issue is not yet fixed still it's slow to pick mov files. Below log are shown in console. any other solution for this?

***** Picker:<PHPickerViewController: 0x1208d7b00> didFinishPicking:(
    "<PHPickerResult: 0x283b466a0>"
)
* -[FilePickerPlugin picker:didFinishPicking:] [Line 381]


Thanks in Advance

**Comment by neelansh-creatorstack** [comment]
The same .mov file when selected in the using the FileType.any( i.e from the device explorer), it is selected within 1 sec, while when passing the file type as FileType.video and picking the same video from the Video Gallery, it takes around 40 sec.
cc: @miguelpruivo 

I've tried this with several files. The more the fileSize, the more the time taken when picking from Gallery.

Is the `withData` property somehow `true` when using the fileType as Video?

**Comment by renanyoy** [comment]
I have same problem on mac, i just need the path, so all data, stream flags set to false, it can take 20s for 40 files selected..
